### ISSUE TITLE
FR: TrueNAS CORE (FreeBSD) support

### ISSUE THREAD

**Issue by SnoFox** [tailscale/tailscale#3694]
### What are you trying to do?

I'd like to directly link a TrueNAS CORE client to the network.

### How should we solve this?

I would like to see this as part of the services UI alongside other TrueNAS services, such as Wireguard or iSCSI, for a one-click install. How you'd OAuth afterwards isn't too important to the workflow as long as you disable key expiry afterwards.

### What is the impact of not solving this?

Current workarounds involve using a separate server in the local network to the NAS or VM on the TrueNAS box acting as a subnet router, creating extra failure points/services to maintain.

### Anything else?

iXsystems has an open feature request for this as well: https://jira.ixsystems.com/browse/NAS-110540, but it seems unlikely they'll be working on it soon.

I've decided to open this separately from https://github.com/tailscale/tailscale/issues/3097, a request for the Debian-based TrueNAS SCALE, verses TrueNAS CORE (née FreeNAS), a BSD-based OS.

**Comment by bradfitz** [comment]
So for CORE (FreeBSD) they're call plugins (https://www.truenas.com/apps/#Plugins) and SCALE (Linux) they're called Apps (https://www.truenas.com/apps/#Apps)?

Looks like the CORE docs/guide is at https://www.truenas.com/docs/core/applications/plugins/createplugin/ which promisingly lists `allow_tun` and `allow_raw_sockets` and `nat`, etc.

I don't imagine Tailscale will work on this anytime soon (unless we a particular customer requiring it perhaps), but thanks for filing the tracking bug to gather notes.

**Comment by SnoFox** [comment]
This seems to be the terminology they use - I only became aware of SCALE myself when filing this FR. With my testing, it appears CORE runs plugins from a BSD Jail, while SCALE runs apps in a docker-backed Kubernetes cluster.

That said, services seem to be built into the OS - the latest release brings both and OpenVPN client and server in (https://www.truenas.com/docs/core/services/openvpn/) with nice configuration UIs, while plugins can be community created, which OpenVPN Server previously was.

Regardless, plugin or service, it'd be nice to shut down extra subnet router as a workaround.

**Comment by tzneal** [comment]
This is a TrueNAS core plugin I messed with a while back, haven't really looked at it since though: https://github.com/tzneal/iocage-plugin-tailscale

If you have a /mnt/Taildrop directory mounted inside the jail, taildrop'd files should go there.

**Comment by DentonGentry** [comment]
https://truecharts.org/docs/charts/stable/tailscale/How-To-Guide/ provides a TrueNAS add-on for Tailscale connectivity.

**Comment by zacwest** [comment]
That is TrueNAS Scale (Linux) using containers, not Core (BSD).

**Comment by bradfitz** [comment]
Given that TrueNAS CORE seems to be a dead end with them pushing users to TrueNAS SCALE, I don't see Tailscale working on this anytime soon.

If anybody wants this, https://github.com/tailscale/tailscale/issues/3694#issuecomment-1009579731 above looks like a good starting point.

If somebody wants to revive this, feel free to advertise your work here.

**Comment by SnoFox** [comment]
I agree that CORE is on a slow path to complete deprecation and this FR should be closed. 

I’ve personally sidegraded to SCALE as well. Still using a subnet router due to difficulty running Tailscale client inside a container for the host OS, but it still obviates the need for this request.
